server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java

```diff
@@ -0,0 +1,424 @@
+        IndexMetadata oldestIndexMeta = metadata.getIndexSafe(oldestBackingIndex);
+
+        return dataStream.getAutoShardingEvent() == null
+            // without a pre-existing auto sharding event we wait until the oldest index has been created longer than the decrease_shards
+            // cool down period "ago" so we don't immediately after a data stream is created, reduce the number of shards
+            ? TimeValue.timeValueMillis(
+                Math.max(0L, oldestIndexMeta.getCreationDate() + reduceShardsCooldown.millis() - nowSupplier.getAsLong())
+            )
+            : TimeValue.timeValueMillis(
+                Math.max(
+                    0L,
+                    reduceShardsCooldown.millis() - dataStream.getAutoShardingEvent()
+                        .getTimeSinceLastAutoShardingEvent(nowSupplier)
+                        .millis()
+                )
+            );
+    }
+
+    private AutoShardingResult getDecreaseShardsResult(
+        Metadata metadata,
+        DataStream dataStream,
+        double writeIndexLoad,
+        LongSupplier nowSupplier,
+        IndexMetadata writeIndex,
+        TimeValue remainingReduceShardsCooldown
+    ) {
+        double maxIndexLoadWithinCoolingPeriod = getMaxIndexLoadWithinCoolingPeriod(
+            metadata,
+            dataStream,
+            writeIndexLoad,
+            reduceShardsCooldown,
+            nowSupplier
+        );
+
+        logger.debug(
+            "calculating the optimal number of shards for a potential decrease in number of shards for data stream [{}] with the"
+                + " max indexing load [{}] over the decrease shards cool down period",
+            dataStream.getName(),
+            maxIndexLoadWithinCoolingPeriod
+        );
+        long optimalShardCount = computeOptimalNumberOfShards(minWriteThreads, maxWriteThreads, maxIndexLoadWithinCoolingPeriod);
+        if (optimalShardCount < writeIndex.getNumberOfShards()) {
+            logger.debug(
+                "data stream autosharding service recommends decreasing the number of shards from [{}] to [{}] after [{}] cooldown for "
+                    + "data stream [{}]",
+                writeIndex.getNumberOfShards(),
+                optimalShardCount,
+                remainingReduceShardsCooldown,
+                dataStream.getName()
+            );
+
+            // we should reduce the number of shards
+            return new AutoShardingResult(
+                AutoShardingType.DECREASE_SHARDS,
+                writeIndex.getNumberOfShards(),
+                Math.toIntExact(optimalShardCount),
+                remainingReduceShardsCooldown,
+                maxIndexLoadWithinCoolingPeriod
+            );
+        }
+
+        return new AutoShardingResult(
+            AutoShardingType.NO_CHANGE_REQUIRED,
+            writeIndex.getNumberOfShards(),
+            writeIndex.getNumberOfShards(),
+            TimeValue.ZERO,
```

Comment:
Same comment here about the weirdness of passing 0 when we could maybe expand the enum and give more info (including a calculation)